Include the Mastodon URL in a link element on user page (#2046)

nud opened the pull request:
We add a `<link rel="me" href="https://mastodon.example/@me">` URL in  the `<head>` when the user configures their Mastodon URL.
    
We only accept URLs (not `@me@mastodon.example`-style identifiers) for    the sake of simplicity, because it's both the most universal format and    the one required for the "verification" feature. A placeholder is used     to convey that fact.
    
For the record, at this time Mastodon requires a `<a>` or `<link>`     element with `rel="me"` and an `href` attribute pointing to the     `https://` URL of the Mastodon account. I opted for the `<link>` element     because it didn't involve modifying the web interface any further. See     https://docs.joinmastodon.org/user/profile/#verification for details.
    
https://linuxfr.org/suivi/faire-valider-sa-page-perso-linuxfr-org-sur-mastodon

I also fixed the formatting of command lines in Docker.md while I was at it.

## Updated options

![image](https://user-images.githubusercontent.com/10519/210018871-bb31eb2d-e92a-4386-b12b-3aa217d567b2.png)

## `<head>`

![image](https://user-images.githubusercontent.com/10519/210018910-9f42a054-d2f3-47b2-8721-199538ff4704.png)

Trim commented:
Thanks for this PR, it's nice.

I've added a comment on the issue report: https://linuxfr.org/suivi/faire-valider-sa-page-perso-linuxfr-org-sur-mastodon#comment-1910962

For the part concerning this pull request, it would be nice to add these two things:

First a string after the Mastodon input saying something like below to discover this functionality.

> Après avoir enregisté le lien de votre compte Mastodon, vous pourrez utiliser le lien https://linuxfr.org/users/xxx pour faire vérifier votre identité Mastodon via LinuxFr.

Second a link to the user Mastodon account on each user content as there's already this functionality for personal website and xmpp account:

> (site Web personnel, adresse XMPP, compte Mastodon)

It would be nice to discover new Mastodon account to follow.

flzara commented:
> Thanks for this PR, it's nice.
> 
> I've added a comment on the issue report: https://linuxfr.org/suivi/faire-valider-sa-page-perso-linuxfr-org-sur-mastodon#comment-1910962

Thanx, I added some comments just below yours: https://linuxfr.org/suivi/faire-valider-sa-page-perso-linuxfr-org-sur-mastodon#comment-1910974

> 
> For the part concerning this pull request, it would be nice to add these two things:
> 
> First a string after the Mastodon input saying something like below to discover this functionality.
> 
> > Après avoir enregisté le lien de votre compte Mastodon, vous pourrez utiliser le lien https://linuxfr.org/users/xxx pour [faire vérifier votre identité Mastodon](https://docs.joinmastodon.org/user/profile/#verification) via LinuxFr.

May I suggest a little modification at the end : 

> Après avoir enregisté le lien de votre compte Mastodon, vous pourrez utiliser le lien https://linuxfr.org/users/xxx pour [prouver que celui-ci appartient bien au même utilisateur que sur LinuxFr.org](https://docs.joinmastodon.org/user/profile/#verification).

nud commented:
@Trim:

> Second a link to the user Mastodon account on each user content as there's already this functionality for personal website and xmpp account:

I didn't do it on purpose, but I suppose it can be added if you insist.

For the record I'm not sure adding all the links to someone on each comment is the way to go (I suppose having a proper profile page would be better IMO) because if someone adds for e.g. Matrix or Signal support now that XMPP is pretty much dead, will you add these too? Knowing that removing a link (e.g. XMPP) without having people complaining will be pretty much impossible :wink:

nud commented:
Obviously I didn't test the actual verification process...

nud commented:
@Trim @flzara Requested changes have been implemented.

Trim commented:
> Obviously I didn't test the actual verification process...

I've checked it with my new mastodon account. It seems to work well.

I've a glitch which hide it after a little moment on my Mastodon instance, but I'm sure it's my fresh Mastodon install which is missing some feature, as it's my first Mastodon install.

Trim commented on app/helpers/users_helper.rb:
```diff
@@ -35,9 +35,12 @@ def homesite_link(user)
   end
 
   def jabber_link(user)
-    return unless current_account
```
Well, I think we should keep this condition for `jabber_link`: this condition hides the link for LinuxFr visitors (anonymous web user).

It avoids to disclose too easily XMPP addresses and so it should avoid some spam on the XMPP accounts.

It's not needed for `mastodon_link`, because Mastodon accounts are already public, isn't it ?

nud replied:
I removed the offending line from the commit.

nud commented:
I got inspired by @Oumph's latest merged commit and added statistics for Mastodon domains while I was at it.

This commit series talks about Mastodon all over the place but in truth some of it would work with other ActivityPub software (Pleroma comes to mind, but there are lots of others). Although as far as I know the verification process is Mastodon-specific, maybe the wording will need to be adapted if we show that URI in the web interface. I honestly have no idea what the proper word is / will be when the dust settles.

nud commented:
In the same spirit as https://github.com/linuxfrorg/linuxfr.org/pull/355 I updated this MR to use the `http_url` validator. I also changed the field type in preferences to be a "url" field rather than a "text" field (similar to the homesite field)

Trim reviewed (approved):
Looks good, thanks for the improvements !

@Oumph  can merge and publish :+1: 

PS: I don't worry about the fact we currently only allow https link for Mastodon profile (instead of the user name), because the [Mastodon documentation](https://docs.joinmastodon.org/user/profile/#verification) enforces the use the https profile link to verify the profile:

> the href attribute on one of those elements is equal to the URL for your Mastodon profile

Oumph commented on app/models/user.rb:
```diff
@@ -33,6 +34,8 @@ class User < ActiveRecord::Base
   validates :homesite, length: { maximum: 100, message: "L’adresse du site Web personnel est trop longue" }
   validates :name, length: { maximum: 40, message: "Le nom affiché est trop long" }
   validates :jabber_id, length: { maximum: 32, message: "L’adresse XMPP est trop longue" }
+  validates :mastodon_url, http_url: { protocols: ["https"], message: "L’adresse du compte Mastodon n’est pas une URL valide" },
```
```suggestion
  validates :mastodon_url, http_url: { protocols: ["https"], message: "L’adresse du compte Mastodon n’est pas valide" },
```

Oumph commented on app/views/devise/registrations/_preferences.html.haml:
```diff
@@ -11,6 +11,12 @@
+      = u.label :mastodon_url, 'Adresse du compte Mastodon'
+      = u.url_field :mastodon_url, placeholder: "https://mastodon.example/@me", maxlength: 255
+      %span.help
+        Après avoir enregistré l'adresse de votre compte Mastodon, vous pourrez utiliser le lien <kbd>https://linuxfr.org/users/#{current_user.login}</kbd> pour
```
```suggestion
        Après avoir enregistré l’adresse de votre compte Mastodon, vous pourrez utiliser le lien <kbd>https://linuxfr.org/users/#{current_user.login}</kbd> pour
```

flzara commented:
Merci :)
